src/interface/hooks/use-issue-license.js

```diff
@@ -0,0 +1,155 @@
+          proposal.connection_id,
+          proposal.by_format.pres_proposal.indy,
+          moment().format('YYYY-MM-DD'),
+          (data) => {
+            resolve(data)
+          }
+        )
+      }),
+    [postRequest]
+  )
+
+  const deleteRecordP = useCallback(
+    (origin, presExId) =>
+      new Promise((resolve) => {
+        deleteRecord(origin, presExId, () => {
+          resolve()
+        })
+      }),
+    [deleteRecord]
+  )
+
+  const getRecordP = useCallback(
+    (origin, presExId) =>
+      new Promise((resolve) => {
+        getRecord(origin, presExId, (data) => {
+          resolve(data)
+        })
+      }),
+    [getRecord]
+  )
+
+  const postCredOfferP = useCallback(
+    (origin, verifiedRecord, licenseCredDefId) =>
+      new Promise((resolve) => {
+        const attr =
+          verifiedRecord.by_format.pres.indy.requested_proof.revealed_attrs
+        const expiry = moment().add(2, 'years').format('YYYYMMDD')
+
+        postCredOffer(
+          origin,
+          verifiedRecord.connection_id,
+          licenseCredDefId,
+          attr['0_id_uuid'].raw,
+          attr['0_type_uuid'].raw,
+          expiry,
+          verifiedRecord.pres_request.comment,
+          (data) => {
+            resolve(data)
+          }
+        )
+      }),
+    [postCredOffer]
+  )
+
+  const waitForVerifiedRecord = useCallback(
+    async (origin, requestPresExId) => {
+      let i = 0
+      const retryCount = 10
+      for (; i < retryCount; i++) {
+        const record = await getRecordP(origin, requestPresExId)
+        if (record?.state === 'done') {
+          return record
+        }
+        await delay(3000)
+      }
+      if (i === retryCount) {
+        setTimeoutError('timeout')
+        return null
+      }
+    },
+    [getRecordP]
+  )
+
+  const onStartFetch = useCallback(
+    async (origin, proposal, licenseCredDefId) => {
+      if (!inflight.has(proposal.pres_ex_id)) {
+        setInflight((prev) => new Set(prev.add(proposal.pres_ex_id)))
+
+        const requestPresExId = await postRequestP(origin, proposal)
+        await deleteRecordP(origin, proposal.pres_ex_id)
+        const verified = await waitForVerifiedRecord(origin, requestPresExId)
+        if (!verified) return
+        const licenseId = await postCredOfferP(
+          origin,
+          verified,
+          licenseCredDefId
+        )
+        await deleteRecordP(origin, verified.pres_ex_id)
+
+        setInflight(
+          (prev) => new Set([...prev].filter((id) => id != proposal.pres_ex_id))
+        )
+        setStatus('done')
+        return licenseId
+      }
+    },
+    [
+      inflight,
+      postRequestP,
+      deleteRecordP,
+      waitForVerifiedRecord,
+      postCredOfferP,
+    ]
+  )
+
+  if (
+    statusRequest === 'error' ||
+    statusRecords === 'error' ||
+    statusIssueCred === 'error' ||
+    statusDelete === 'error'
+  ) {
+    setStatus('error')
+  }
+
+  return [
+    status,
+    errorRequest ||
```

Comment:
It would be nice to refactor hooks to use a shared object for fetching state along with error/success

Reply:
Maybe. Normally I'd expect a context component higher up the stack (like Auth0 does). Then that would be able to report API success error generically. I think any sort of major refactor like that has to be out of scope for this work

Reply:
yea, i like your idea